File: android_webview/java/src/org/chromium/android_webview/AwCrashyClassUtils.java
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.android_webview;

import androidx.annotation.VisibleForTesting;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.android_webview.common.AwFeatures;
import org.chromium.android_webview.common.AwSwitches;
import org.chromium.base.CommandLine;

/** A helper class for testing related to Java and Native crashes. */
@JNINamespace("android_webview")
public final class AwCrashyClassUtils {

    public static void maybeCrashIfEnabled() {
        if (shouldCrashJava()) {
            throw new RuntimeException("WebView Forced Java Crash for WebView Browser Process");
        } else if (shouldCrashNative()) {
            AwCrashyClassUtilsJni.get().crashInNative();
        }
    }

    @VisibleForTesting
    public static boolean shouldCrashJava() {
        return AwFeatureMap.isEnabled(AwFeatures.WEBVIEW_ENABLE_CRASH)
                && CommandLine.getInstance().hasSwitch(AwSwitches.WEBVIEW_FORCE_CRASH_JAVA);
    }

    public static boolean shouldCrashNative() {
        return AwFeatureMap.isEnabled(AwFeatures.WEBVIEW_ENABLE_CRASH)
                && CommandLine.getInstance().hasSwitch(AwSwitches.WEBVIEW_FORCE_CRASH_NATIVE);
    }

    // Do not instantiate this class.
    private AwCrashyClassUtils() {}

    @NativeMethods
    interface Natives {
        void crashInNative();
    }
}

File: android_webview/browser/aw_crashy_class_utils.cc
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <ostream>

#include "base/compiler_specific.h"
#include "base/logging.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "android_webview/browser_jni_headers/AwCrashyClassUtils_jni.h"

namespace android_webview {

// Do not inline this function. This is meant to simulate a crash like we would
// see in a real incident. Real crashes will generally have at least 1
// non-inlined stack frame, so we force this to happen in this case using both
// the NOINLINE and NOOPT annotations (for some reason, NOINLINE is not enough).
NOINLINE NOOPT void ThisFunctionWillCrash() {
  LOG(FATAL) << "WebView Forced Native Crash for WebView Browser Process";
}

static void JNI_AwCrashyClassUtils_CrashInNative(JNIEnv* env) {
  ThisFunctionWillCrash();
}

}  // namespace android_webview
